data-raw/process_ecmwf_mars_gribs_to_cogs.py

```diff
@@ -0,0 +1,131 @@
+import xarray as xr
+from affine import Affine
+from rasterio.crs import CRS
+from scipy.interpolate import griddata
+
+from src.blob_utils import (
+    download_file,
+    list_blobs,
+    load_env_vars,
+    upload_file,
+)
+
+sas_token, container_name, storage_account = load_env_vars()
+
+
+grib_files = list_blobs(
+    sas_token=sas_token,
+    container_name=container_name,
+    storage_account=storage_account,
+    name_starts_with="ds-aa-afg-drought/raw",
+)
+
+grib_files = [file for file in grib_files if file.endswith(".grib")]
+grib_file_samp = grib_files[:1]
+# Loop through each blob path
+for blob_name in grib_file_samp:
+    print(f"name: {os.path.basename(blob_name)}")
+
+    # Create a temporary file
+    with tempfile.TemporaryDirectory() as td:
+        temp_base_path = os.path.basename(blob_name)
+        td = Path(td)
+        tf = td / temp_base_path
+        # Download the file
+        download_file(
+            sas_token=sas_token,
+            container_name=container_name,
+            storage_account=storage_account,
+            blob_path=blob_name,
+            local_file_path=tf,
+        )
+
+        # Open the file with xarray
+        ds = xr.open_dataset(
+            tf,
+            engine="cfgrib",
+            drop_variables=["surface", "values"],
+            backend_kwargs=dict(time_dims=("time", "forecastMonth")),
+        )
+
+        ds["longitude"] = (ds["longitude"] + 180) % 360 - 180
+        ds["precip"] = (
+            ds["tprate"] * ds["time"].dt.days_in_month * 24 * 3600 * 1000
+        )
+        ds = ds.drop_vars("tprate")
+        da = ds["precip"]
+        da_mean = da.mean(dim=["number"])
+
+        # loop through pub dates & COGS
+        pub_dates = da_mean.time.values
+        forecast_months = da_mean.forecastMonth.values
+
+        # give a CRS
+        da_mean = da_mean.rio.set_crs("EPSG:4326", inplace=True)
+
+        # set up grid
+        rRes = 0.4
+        points = list(zip(da_mean.longitude.values, da_mean.latitude.values))
+        xRange = np.arange(da.longitude.min(), da.longitude.max() + rRes, rRes)
+        yRange = np.arange(da.latitude.min(), da.latitude.max() + rRes, rRes)
+
+        gridX, gridY = np.meshgrid(xRange, yRange)
+
+        # define transform and CRS
+        transform = Affine.translation(
+            gridX[0][0] - rRes / 2, gridY[0][0] - rRes / 2
+        ) * Affine.scale(rRes, rRes)
+        rasterCrs = CRS.from_epsg(4326)
+
+        for pub_date in pub_dates:
+            da_filt = da_mean.sel(time=pub_date)
+            time_temp = pd.to_datetime(str(pub_date))
+            time_str_temp = time_temp.strftime("%Y-%m-%d")
+            print(time_str_temp)
+
+            # create forecast month vector to loop through
+            fms = da_filt.forecastMonth.values
+            for i, fm_temp in enumerate(fms):
+                print(fm_temp)
+                da_filt_lt = da_filt.sel(forecastMonth=fm_temp)
+                grid_ecmwf = griddata(
```

Comment:
So I've tried a bunch of versions of this, but always end up with the same issue where the top and right sides, and bottom left corner are missing. This seems to be due to the irregular grid of the MARS output, which looks roughly like this:
![image](https://github.com/OCHA-DAP/ds-aa-afg-drought/assets/6000042/e38ad709-8774-42be-bc3d-fbb40746df56)

Since it seems like MARS re-grids based on your query bbox, it seems to have good coverage for at least the first longitude value (see the regular bars on the left side). But then things get messier, probably due to some transformation it applies to its raw data as it generates the output. Then when you interpolate with `griddata()`, it doesn't get the final values because it hasn't been told to extrapolate (its default behaviour is to just set `nan` outside where it has data). 

So the way I see it there are two options to fix this:
1. Make the bounding box bigger, say by 1 degree North and East. This would be better because then you're always interpolating.
2. Just fill the outside values with extrapolation, if you don't want to re-download everything.

For option 2: if you follow this line (`grid_ecmwf = griddata(...)`) with:

`interp = NearestNDInterpolator(points, da_filt_lt.values)`
`extrapolated_values = interp(gridX, gridY)`
`grid_ecmwf = np.where(np.isnan(grid_ecmwf), extrapolated_values, grid_ecmwf)`

then you will basically fill those outside values by taking the nearest value from the original array, I believe. But this isn't ideal. I think just expanding the bounding box is better.

Reply:
wow impressive sleuthing!

I find it bizzarre that MARS returns an irregular grid, but that makes sense that the interpolation attempt I do later with griddata() would return the NA values -- so at least we know why that is - nice!

In this case a buffer is built into the bbox so I think there is not actually a change required (see screenshot below). However, this is certainly something to be aware of. I am curious to see what the grid will look like when we try to get it globally 

![image](https://github.com/OCHA-DAP/ds-aa-afg-drought/assets/23731112/348ac53b-23fe-45f9-91f6-4a357d5a6e1f)